src/nvim/tui/tui.c

```diff
@@ -527,7 +529,13 @@ static void tui_flush(UI *ui)
   }
 
   unibi_goto(ui, data->row, data->col);
+
+  if (!data->busy) {
+    unibi_out(ui, unibi_cursor_normal);
+  }
+
   flush_buf(ui);
+  unibi_out(ui, unibi_cursor_invisible);
```

Comment:
Can you try this patch?

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 610419d..c25ec28 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -46,6 +46,9 @@ typedef struct {
   int old_height;
   bool can_use_terminal_scroll;
   bool busy;
+  // true if there was an update that requires temporaril switching the cursor
+  // off
+  bool require_invisible_cursor;
   HlAttrs attrs, print_attrs;
   Cell **screen;
   struct {
@@ -342,16 +345,21 @@ static void tui_cursor_goto(UI *ui, int row, int col)
   data->row = row;
   data->col = col;
   unibi_goto(ui, row, col);
+  data->require_invisible_cursor = true;
 }

 static void tui_busy_start(UI *ui)
 {
-  ((TUIData *)ui->data)->busy = true;
+  TUIData *data = ui->data;
+  data->busy = true;
+  data->require_invisible_cursor = true;
 }

 static void tui_busy_stop(UI *ui)
 {
-  ((TUIData *)ui->data)->busy = false;
+  TUIData *data = ui->data;
+  data->busy = false;
+  data->require_invisible_cursor = true;
 }

 static void tui_mouse_on(UI *ui)
@@ -475,6 +483,7 @@ static void tui_scroll(UI *ui, int count)
     invalidate(ui, data->scroll_region.top, data->scroll_region.bot,
         data->scroll_region.left, data->scroll_region.right);
   }
+  data->require_invisible_cursor = true;
 }

 static void tui_highlight_set(UI *ui, HlAttrs attrs)
@@ -529,13 +538,7 @@ static void tui_flush(UI *ui)
   }

   unibi_goto(ui, data->row, data->col);
-
-  if (!data->busy) {
-    unibi_out(ui, unibi_cursor_normal);
-  }
-
   flush_buf(ui);
-  unibi_out(ui, unibi_cursor_invisible);
 }

 static void tui_suspend(UI *ui)
@@ -789,12 +792,27 @@ end:

 static void flush_buf(UI *ui)
 {
-  static uv_write_t req;
-  static uv_buf_t buf;
+  uv_write_t req;
+  uv_buf_t buf[2];
+  int buf_index = 0;
+  char invis_buf[64];
   TUIData *data = ui->data;
-  buf.base = data->buf;
-  buf.len = data->bufpos;
-  uv_write(&req, (uv_stream_t *)&data->output_handle, &buf, 1, NULL);
+
+  if (data->require_invisible_cursor) {
+    if (!data->busy) {
+      unibi_out(ui, unibi_cursor_normal);
+    }
+
+    const char *str = unibi_get_str(data->ut, unibi_cursor_invisible);
+    buf[0].base = invis_buf;
+    buf[0].len = unibi_run(str, data->params, invis_buf, sizeof(invis_buf));
+    buf_index++;
+    data->require_invisible_cursor = false;
+  }
+
+  buf[buf_index].base = data->buf;
+  buf[buf_index].len = data->bufpos;
+  uv_write(&req, (uv_stream_t *)&data->output_handle, buf + buf_index, 1, NULL);
   uv_run(data->write_loop, UV_RUN_DEFAULT);
   data->bufpos = 0;
 }
```

I didn't notice any difference in performance and it seems to complicate the code too much. Whats your opinion?